fendermint/vm/interpreter/src/chain.rs

```diff
@@ -89,32 +97,65 @@ where
             CheckpointPoolItem::BottomUp(ckpt) => ChainMessage::Ipc(IpcMessage::BottomUpExec(ckpt)),
         });
 
+        // Prepare top down proposals
+        match atomically_or_err::<_, fendermint_vm_topdown::Error, _>(|| {
+            finality_provider.next_proposal()
+        })
+        .await?
+        {
+            None => {}
+            Some(proposal) => {
+                msgs.push(ChainMessage::Ipc(IpcMessage::TopDownExec(ParentFinality {
+                    height: proposal.height as ChainEpoch,
+                    block_hash: proposal.block_hash,
+                })))
+            }
+        }
+
         // Append at the end - if we run out of block space, these are going to be reproposed in the next block.
         msgs.extend(ckpts);
         Ok(msgs)
     }
 
     /// Perform finality checks on top-down transactions and availability checks on bottom-up transactions.
-    async fn process(&self, pool: Self::State, msgs: Vec<Self::Message>) -> anyhow::Result<bool> {
+    async fn process(&self, state: Self::State, msgs: Vec<Self::Message>) -> anyhow::Result<bool> {
```

Comment:
```suggestion
    async fn process(&self, (pool, provider): Self::State, msgs: Vec<Self::Message>) -> anyhow::Result<bool> {
```
I think this works and would make the rest of the code more readable. `state.0.get_status` is a head scracher.